pkg/apis/pipeline/v1beta1/pipeline_types.go

```diff
@@ -135,6 +139,35 @@ type PipelineTask struct {
 	Timeout *metav1.Duration `json:"timeout,omitempty"`
 }
 
+// FinalPipelineTask defines a final task in a Pipeline, passing inputs from both
+// Params and from the output of non-final tasks.
+type FinalPipelineTask struct {
```

Comment:
What are the differences between this struct and the existing PipelineStruct task? can we use the same struct but apply different validation?

Reply:
A bit more context: im thinking that if we use the exact same PipelineTask, but have some internal notion of whether the PipelineTask is a final task or not, they could share a lot of the same execution logic (and then youd get things like timeouts and retries for free like @skaegi is pointing out, in his words "treat it no differently from a regular Task but the reordering rules are different"

another way to do it in Go would be to have types such as:
- PipelineTask <-- which has the bare minimum fields shared by both PipelineTask (today) and FinalPipelineTask
- DAGPiplineTask (or something) <-- which also includes runAfter, etc.

Reply:
Yup, certainly we can, the idea to create this new struct was to keep the implementation clean vs adding `if finally` at the validation level which would make it even more complicated when we add `except` 😢  Also, one more reason was to enforce the fact that `tasks` and `finally` tasks are actually different from the perspective of specifying dependencies - no `runAfter` supported under `finally`. 

After having this discussed in the API workflow as well, I will change it to use the same `PipelineTask` struct.

Reply:
> another way to do it in Go would be to have types such as:
> 
> * PipelineTask <-- which has the bare minimum fields shared by both PipelineTask (today) and FinalPipelineTask
> * DAGPiplineTask (or something) <-- which also includes runAfter, etc.

I had thought about this but discarded the idea as it resulted in embedding structure into structure (I am fine either ways, dont have strong preference):

```
type PipelineTask struct {
    name string
}

type DAGPipelineTask struct {
    PipelineTask  //embedding
    runAfter string
}
``` 

This results in changing all of the unit tests from:

```
Tasks: []PipelineTask{{						
		Name: "task1",
```
to

```
Tasks: []DAGPipelineTask{{
   PipelineTask:
     PipelineTask{
	Name: "task1",
```

Referencing those fields (Name and etc) would still remain the same.

Reply:
Whatever type we decide to go with, should not impact execution logic since we are adding an additional field `[FinalPipelineTask](https://github.com/tektoncd/pipeline/pull/2437/files#diff-4e301064fef70f3c0db810267c0959f2R73)` to maintain the list of final tasks which would not interfere with non-final tasks.

Reply:
hey @bobcatfish I have listed pros and cons of refactoring `PipelineTask` [here](https://gist.github.com/pritidesai/54ea8efa60692cdc2582cfc60eb12a8b) 😱